### ISSUE TITLE
CFP: Include more bundled envoy extensions

### ISSUE THREAD

**Issue by rauanmayemir** [cilium/cilium#27008]
## Cilium Feature Proposal

We are currently migrating to Cilium for both CNI and service mesh after having been using Istio for several years now. Istio had been great for us, but upon critical review it's become clear that we've been paying the price of running workloads behind sidecars and what feels like an overhead for about 5% of what we actually needed and used from istio.

When we started adopting Cilium, we almost gave up because Cilium turned out to be severely restricting us from using istio in a usual way. But then fortunately we discovered that Cilium has started venturing into service mesh scene and I believe it's shaping up in a very coherent way.

As per [documentation](https://docs.cilium.io/en/stable/network/servicemesh/l7-traffic-management/#supported-envoy-extension-resource-types), cilium Envoy build's support for extensions is rather limited. It might have made sense for cilium when used strictly for policy enforcements, but for service mesh use and general L7 traffic management I believe more envoy extensions should be bundled, or at least there should be a way to plug in less popular extensions more easily.

We've also jumped ahead and started using the envoy daemonset ran separately from cilium agent, which I believe should be the default way. Considering this, a dedicated envoy should be able to have more extensions without interfering with the agent.

While I was writing this FR, I've noticed there are now more built-in filters available which means I've just come to the party a little bit too early. 😄

Based on my own experience, I would like to suggest the following extensions in the order of importance from most to least (excluding those that are already bundled as of recently):
1. `envoy.filters.http.grpc_json_transcoder` and `envoy.filters.http.grpc_web` - we build all our APIs in gRPC and leverage envoy for easy protocol translation. It has proven to work without a hitch, that speaks for its stability
2. `envoy.filters.http.connect_grpc_bridge` - new emerging gRPC conversion protocol that is helping to solve some of the gRPC shortcomings (lighter infra)
3. `envoy.filters.http.compressor` - browser output compression, could have solved this in the upstream, but our judgement was to put this on envoy side (because it was the last one to touch the response and we might have needed to do something with the response)
4. `envoy.filters.http.buffer` - this is to simply be able to limit the inbound request size, but there are other uses 
5. `envoy.filters.http.header_to_metadata` - very useful, but could be solved in the upstream
6. `envoy.matching.inputs.request_headers` - I don't even see this in the config. there is an issue in Envoy where max request size limit breaks websocket, so we use this alongside `envoy.extensions.common.matching.v3.ExtensionWithMatcher` to skip requests with `Upgrade: websocket` and not apply the buffer filter

I hope people will see this and start adding their own suggestions. I've seen people ask Lua filter, we've been wary of using scripting in the LB and usually handled things in the upstream, but I see how valuable it could be in places where `header_to_metadata` couldn't handle more complex cases.

**Comment by rauanmayemir** [comment]
P.S: I've also seen the lack of distributed tracing support, this could use some polishing.

**Comment by youngnick** [comment]
Thanks for this CFP @rauanmayemir. When you say you'd like these available, what do you mean? Do you want them just available in the proxy build, or would you like the functionality exposed in config too?

I guess another way to ask this is: How are you expecting to use these filters?

**Comment by rauanmayemir** [comment]
@youngnick Oh, I didn't realise being able to create those filters in the config weren't same as having them bundled in envoy. If they are not exposed in config, how can they be used inside proxy build?

Good question, how can I use them? I was thinking I would add a `CiliumEnvoyConfig`, but I was kinda expecting to have something similar to what istio does: create a CRD with my filter and a config where I want it placed.

**Comment by youngnick** [comment]
Thanks, that's the reason I asked! 😊

You could use them with a `CiliumEnvoyConfig`, but right now the user experience of doing that for ingress isn't great (you'll have to ensure that the CEC matches the names Cilium will give in the autogenerated CEC, and make sure you don't cause any conflicts because CECs don't have great conflict resolution at the moment, and so on.)

It sounds like a lot of the filters you're mentioning could be addressed by Gateway API Route filters, but upstream hasn't built that (I'm a maintainer there, so that's on me as well, hah).

I agree that all the stuff you're saying here _should_ be something doable, but it might take a little bit to build the config so it's _right_, and not hacked together.

**Comment by rauanmayemir** [comment]
Right, I think it would be perfectly fine to get by with carefully stitching CESs until we have a solid approach.

**Comment by rauanmayemir** [comment]
Has anything changed with Cilium's plans for service mesh?

If this is not something feasible in the short-term, I would consider running my own envoy behind cilium's global proxy and have all the filters in the userspace. Although it is going to be extra pain.

**Comment by youngnick** [comment]
Nothing has changed that I'm aware of - our plan is, broadly:
- Ingress objects used for config will stick around forever (as per the upstream policies).
- Gateway API for ingress will also be supported for greater control.
- For east-west layer 7 routing, you can use Layer 7 network policies for now, but we will also be implementing the GAMMA spec for binding `HTTPRoute` objects to `Service` objects.

Again, we can, will, and do add extensions to the Cilium Envoy on a case-by-case basis, but because `CiliumEnvoyConfig` has minimal (basically no) conflict detection, the more complex you make them, and the more filters you use, the more likely for something to go wrong in a way that will be difficult to troubleshoot. That's why we don't just give you all the filters - we want to ensure that the ones we add work well together and add functionality even without having to delve deep into Envoy's config with `CiliumEnvoyConfig`.

I appreciate you asking for these filters, but we need to take the time to make sure we are doing this right - and we have not had the time to dedicate to this.

If you want to prepare a PR to add this yourself, I would recommend building some community consensus for each of the filters you're asking for by bringing this to a community meeting and discussing it, starting a discussion or Google doc, and gathering some support.

**Comment by zengyuxing007** [comment]
For the East-West traffic, perhaps we need to discuss and come up with a core feature list, and then proceed step by step to implement them by using the Gateway API.

**Comment by rauanmayemir** [comment]
@zengyuxing007 I've been reading up on gw api GEPs and it's quite impressive. Too impressive not to wait for them, but too ambitious to be ready soon enough. 😟 

I now realise that even if more envoy extensions will be bundled into cilium's envoy build, using them won't be trivial. Until GW API becomes mature enough to support all the policy attachments (i.e filters and extensions) and until all of them are actually supported in cilium, having insertion points to `CiliumEnvoyConfig` would be a good approach to solve east-west (and even north-south) needs before committing to implementing them. 

I could manually create `CiliumEnvoyConfig`, but that means I will have to manually convert all HTTPRoute, maintain my own LB service, etc. Basically, I want something like `CiliumEnvoyFilter` that patches existing or attaches new filters, routes and listeners to a config maintained by the `CiliumEnvoyConfig`.

**Comment by adalrsjr1** [comment]
I have one more question related to this `CiliumEnvoyConfig`.

I've tried to deploy a WASM plugin using the `CiliumEnvoyConfig` and I couldn't due to a type not recognized:

```
Ignoring invalid CiliumEnvoyConfig JSON (proto:\u00a0(line 1:353): unable to resolve \"type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm\": \"not found\")
```

Reading this thread I understand the reason. However, I have a further question that I want to confirm.

If I bypass the Cilium API and manually configure the Envoy-proxy bootstrap config with my plugin would it works? Is there any known issues when deploying a WASM plugin on Cilium-proxy?

I noted that all extensions related to WASM are enabled in the building script, for example the [`envoy.filters.http.wasm`](https://github.com/cilium/proxy/blob/311e50a272aa108f762689ca85fe6f36a1846a11/envoy_build_config/extensions_build_config.bzl#L164), so, theoretically, there is no blockers that prevent me to configure it. Do I missing anything in this hyphotesis?

**Comment by siegenthalerroger** [comment]
I'm actually wondering the same thing in regards to `proxy-wasm` plugins. Is it possible to load these somehow and then use the CiliumEnvoyConfig to configure these. 

Concretely I'm looking to be able to load the [coraza-waf-wasm plugin](https://github.com/corazawaf/coraza-proxy-wasm) in order to configure the owasp CSR as a WAF for specific ingresses. I guess a cilium-proxy go plugin for coraza waf would also get me there but I'm even less sure of how that would work tbh. Any pointers?

**Comment by adalrsjr1** [comment]
@siegenthalerroger ultimately I could bypass cilium and install my wasm-plugin on Cilium-proxy. However, Cilium-proxy crash raising an error missing `proxy-abi`. 

I don't understand why it's happening as all wasm extensions are enabled on the [`bazel file`](https://github.com/cilium/proxy/blob/main/envoy_build_config/extensions_build_config.bzl).

Anyway, would be great have some feedback from maintainers about this.

**Comment by kevinlmadison** [comment]
Glad to see some discussion about this. I get the desire to prevent people shooting themselves in foot but we've been stuck with istio simply because they allow lua filters. I would much prefer to dispatch with istio entirely and just use cilium.

**Comment by youngnick** [comment]
@adalrsjr1, I've spoken to some other folks who have been able to get WASM plugins working, although of course we don't officially support this at the moment, so I don't know if I can help much more. But I've heard folks say it's _possible_.

@kevinlmadison, I get that a lot of people want to use Lua, but, as I've said before, the problem is that then we (being Cilium) are on the hook for everything that people can do with Lua, including all the footguns. I'd be really interested to hear what you need Lua to be able to do though?

**Comment by adalrsjr1** [comment]
> @adalrsjr1, I've spoken to some other folks who have been able to get WASM plugins working, although of course we don't officially support this at the moment, so I don't know if I can help much more. But I've heard folks say it's _possible_.

@youngnick, thx for following up on this. I'm wondering, did those folks enable any extra extention on cilium-envoy to make it happen? I'm curious, because theoretically I followed the "right" steps to enable my WASM plugin. I even could deploy and see the logs from envoy as expected, but at runtime envoy crashes due to "missing proxy-abi" and this bugs me.

**Comment by youngnick** [comment]
I don't believe any extra extensions were added, no. I'm sorry, it was a while ago and I can't easily ask the person.

**Comment by rauanmayemir** [comment]
Issue wording doesn't reflect that as I created it before understanding how Cilium works, but simply adding stuff into the Envoy image is not enough. We also need a way to be able to apply those extensions to services/routes, preferably via GRPCRoute/HTTPRoute ExtensionRefs.

**Comment by siegenthalerroger** [comment]
@rauanmayemir Cilium has the `CiliumEnvoyConfig` CRD that afaik pretty much lets you use xDS to configure envoy to your hearts content. Have you considered that as GatewayAPI isn't ready for proper extensions yet (once again afaik).

X-Ref: https://github.com/cilium/cilium/issues/25334

**Comment by rauanmayemir** [comment]
Yeah, I know. But I'd rather wait for ExtensionRefs to be supported (or implement it myself wherever possible).

**Comment by peterleezzzZflps** [comment]
I also had problems using the wasm plugin. When I configured cec's wasm plug-in in local mode, my wasm plug-in could run very well. At that time, I used the remote wasm plug-in. It cannot be used. It always reports JSON configuration errors. I am very sure that there is no problem with the configuration because I have tested it with Cilium-envoy:1.29.7. There is no problem at all in remote mode, but I use cec to load remote wasm. The plugin keeps failing. This makes it very inconvenient for me to use.

**Comment by peterleezzzZflps** [comment]
> I also had problems using the wasm plugin. When I configured cec's wasm plug-in in local mode, my wasm plug-in could run very well. At that time, I used the remote wasm plug-in. It cannot be used. It always reports JSON configuration errors. I am very sure that there is no problem with the configuration because I have tested it with Cilium-envoy:1.29.7. There is no problem at all in remote mode, but I use cec to load remote wasm. The plugin keeps failing. This makes it very inconvenient for me to use.


I reported a detailed explanation of this issue on the slack service mesh channel, hoping to get [feedback.](https://cilium.slack.com/archives/C02QKQDTVDX/p1723994804035239)

**Comment by VGerris** [comment]
someone mentioned it working here with a workaround : https://github.com/cilium/cilium/issues/33767

**Comment by YumeAyai** [comment]
> I also had problems using the wasm plugin. When I configured cec's wasm plug-in in local mode, my wasm plug-in could run very well. At that time, I used the remote wasm plug-in. It cannot be used. It always reports JSON configuration errors. I am very sure that there is no problem with the configuration because I have tested it with Cilium-envoy:1.29.7. There is no problem at all in remote mode, but I use cec to load remote wasm. The plugin keeps failing. This makes it very inconvenient for me to use.

@peterleezzzZflps have you already solve it?

**Comment by peterleezzzZflps** [comment]
@YumeAyai
I am no longer actively working in this area, but at the time, the issue
proved unresolvable. My hypothesis was that the problem lay within the
specific Cilium component responsible for downloading remote WASM plugins;
however, when I accessed the container and inspected the downloaded
plugins, they appeared to be intact. Consequently, the issue was never
fully resolved. Ultimately, the workaround I adopted was to store the WASM
plugins within a storage class and mount them to a fixed directory each
time—essentially treating them as local files—which resolved the problem. I
hope this answer proves helpful to you.

Karin ***@***.***> 于2026年2月1日周日 12:14写道：

**Comment by YumeAyai** [comment]
[@peterleezzzZflps](https://github.com/peterleezzzZflps)
I solved it the same way, unfortunately. From what I understand, only Istio supports remote/OCI-based WASM deployment directly. Envoy itself does support remote WASM configuration, so the issue is likely specific to Cilium's Envoy integration—perhaps the build is stripped down or the configuration handling is incomplete.
